cowprotocol/services
# Enable Auto-detection of Token Balance Overrides

# Description

This PR is a follow up to #3148 and the final PR in the stack. It adds configuration and glue code for enabling the automatic token balance override detection introduced in the aforementioned PR.

# Changes

- A new `Arguments` type was created for setting balance override configuration and creating the balance overrides instance.
- The old implementation of the `BalanceOverriding` trait was superseded by a new one that accepts both a static configuration from the `Arguments` as well as allows optionally for auto-detection
- The `Trader` Solidity implementation was changed to only wrap ETH if the trader has sufficient balance. Attempting wrap more than the trader's balance this would cause the simulation to revert without a helpful revert message; and generally means that balance override feature doesn't work for WETH. With the modified implementation, the `Trader.prepareSwap` no longer reverts attempting to wrap more ETH than the balance holds. Now, the `Spardose.requestFunds` will always get called if the `Trader` does not have sufficient `sell_token` balance, and it provides a more helpful message when balance overrides are disabled:
    ```
    2024-12-04T16:09:19.002Z  WARN request{id="2"}:estimator{name="test_quoter"}: shared::price_estimation::trade_verifier: failed verification; returning unverified estimate err=failed to simulate quote

    Caused by:
        simulation reverted Some("revert: trader does not have enough sell token") ...
    ```

## How to Test

An E2E test was modified to include balance override tests for both configured and auto-detected tokens. Additionally, I added a test to ensure that the ETH -> WETH wrapping functionality that previously existed in quote verification simulation did not regress with this change.

## Additional Notes

@MartinquaXD commented that the changes to `Trader` did not make much sense. In order to try and address the comment, I moved some code around. In particular, the `Trader` now calls `requestFunds` from the `Spardose` during the `prepareSwap` call (instead of the `Solver` calling `ensureBalance` for the trader when setting up the settlement). IMO this makes a bit more sense conceptually (it is like the trader had a pre-hook that would provide the required funds for the swap). The contract changes are part of this commit: https://github.com/cowprotocol/services/pull/3149/commits/7518ea917a96d98adcf9b3cb7ce618037ae7bc95.

## Configuration

The balance overriding feature is **disabled by default**. There are two relevant configuration options that need to be changed in order to enable it:
- `QUOTE_TOKEN_BALANCE_OVERRIDES`: a set of `${ADDR}@${SLOT}` configurations for manually enabling token balance overrides for specific tokens (for example, for enabling WETH and USDC on Ethereum Mainnet: `0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2@3,0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48@9`).
- `QUOTE_AUTODETECT_TOKEN_BALANCE_OVERRIDES`: Set to `true` in order to enable auto-detection of balance override slots.
